File: services/go-nethttp/internal/protocol/protocol_test.go
package protocol

import (
	"math"
	"slices"
	"testing"
)

func TestParseStartMessageStrict(t *testing.T) {
	t.Parallel()

	if err := ParseStartMessage([]byte(`{"type":"start"}`)); err != nil {
		t.Fatalf("expected exact start to parse: %v", err)
	}

	cases := [][]byte{
		[]byte(`not-json`),
		[]byte(`{"type":"start","extra":true}`),
		[]byte(`{"type":"partial"}`),
		[]byte(`{"type":"start"} trailing`),
	}
	for _, input := range cases {
		if err := ParseStartMessage(input); err == nil {
			t.Fatalf("expected %q to fail strict start parsing", input)
		}
	}
}

func TestPCMFrameSize(t *testing.T) {
	t.Parallel()

	if !IsPCMFrame(make([]byte, FrameBytes)) {
		t.Fatal("640 byte payload should be a PCM frame")
	}
	if IsPCMFrame(make([]byte, FrameBytes-1)) {
		t.Fatal("short payload should not be a PCM frame")
	}
}

func TestPartialMessageSerializesExactLoadgenSchema(t *testing.T) {
	t.Parallel()

	payload, err := MarshalPartialMessage(PartialMessage{
		Type:              PartialType,
		OldestFrameSeq:    1,
		NewestFrameSeq:    2,
		Frames:            2,
		RMS:               2425.5,
		ZeroCrossings:     7,
		Checksum:          1745339397,
		Samples:           1280,
		Transcript:        "now",
		AudioBytes:        1280,
		CPUPasses:         4,
		ModelDelayMS:      75,
		FlushLatenessMS:   1.23456,
		InflightModelJobs: 0,
	})
	if err != nil {
		t.Fatalf("marshal partial: %v", err)
	}

	object := DecodeObjectForTest(t, payload)
	assertKeys(t, object, []string{
		"type",
		"oldest_frame_seq",
		"newest_frame_seq",
		"frames",
		"rms",
		"zero_crossings",
		"checksum",
		"samples",
		"transcript",
		"audio_bytes",
		"cpu_passes",
		"model_delay_ms",
		"flush_lateness_ms",
		"inflight_model_jobs",
	})
	if object["flush_lateness_ms"] != 1.235 {
		t.Fatalf("flush_lateness_ms = %#v, want 1.235", object["flush_lateness_ms"])
	}
}

func TestErrorMessageSerializesExactDiagnosticSchema(t *testing.T) {
	t.Parallel()

	payload, err := MarshalErrorMessage(ErrorMessage{
		Type:                   ErrorType,
		Stage:                  StageInferenceRequest,
		Kind:                   KindTimeout,
		Message:                "timed out",
		OldestFrameSeq:         1,
		NewestFrameSeq:         2,
		Frames:                 2,
		AudioBytes:             1280,
		OldestAgeMS:            1200.12345,
		NewestAgeMS:            200.12345,
		FlushLatenessMS:        4.4444,
		InferenceElapsedMS:     Float64Ptr(2000.12345),
		InflightGatewayBatches: 1,
		GatewayBufferFrames:    3,
		InferenceStatus:        nil,
		Retryable:              true,
	})
	if err != nil {
		t.Fatalf("marshal error: %v", err)
	}

	object := DecodeObjectForTest(t, payload)
	assertKeys(t, object, []string{
		"type",
		"stage",
		"kind",
		"message",
		"oldest_frame_seq",
		"newest_frame_seq",
		"frames",
		"audio_bytes",
		"oldest_age_ms",
		"newest_age_ms",
		"flush_lateness_ms",
		"inference_elapsed_ms",
		"inflight_gateway_batches",
		"gateway_buffer_frames",
		"inference_status",
		"retryable",
	})
	if object["oldest_age_ms"] != 1200.123 {
		t.Fatalf("oldest_age_ms = %#v, want 1200.123", object["oldest_age_ms"])
	}
	if object["inference_status"] != nil {
		t.Fatalf("inference_status = %#v, want nil", object["inference_status"])
	}
}

func TestDecodePredictionStrictAndValidated(t *testing.T) {
	t.Parallel()

	if _, err := DecodePrediction([]byte(`{"rms":1.5,"zero_crossings":2,"checksum":3,"samples":1280,"transcript":"now","audio_bytes":640}`)); err != nil {
		t.Fatalf("decode valid prediction: %v", err)
	}

	for _, input := range [][]byte{
		[]byte(`{"rms":1.5,"zero_crossings":2,"checksum":3,"samples":1280,"transcript":"now","audio_bytes":640,"extra":true}`),
		[]byte(`{"rms":NaN,"zero_crossings":2,"checksum":3,"samples":1280,"transcript":"now","audio_bytes":640}`),
		[]byte(`{"rms":1.5,"zero_crossings":2,"checksum":3,"samples":0,"transcript":"now","audio_bytes":640}`),
		[]byte(`{"rms":1.5,"zero_crossings":2,"checksum":3,"samples":1280,"transcript":"now","audio_bytes":0}`),
	} {
		if _, err := DecodePrediction(input); err == nil {
			t.Fatalf("expected invalid prediction to fail: %s", input)
		}
	}

	if err := (Prediction{RMS: math.NaN(), Samples: 1, AudioBytes: 1}).Validate(); err == nil {
		t.Fatal("NaN rms should fail prediction validation")
	}
}

func assertKeys(t *testing.T, object map[string]any, expected []string) {
	t.Helper()

	actual := make([]string, 0, len(object))
	for key := range object {
		actual = append(actual, key)
	}
	slices.Sort(actual)
	slices.Sort(expected)
	if !slices.Equal(actual, expected) {
		t.Fatalf("keys = %#v, want %#v", actual, expected)
	}
}

File: services/typescript-bun/tests/protocol.test.ts
import { describe, expect, test } from "bun:test";

import {
  FRAME_BYTES,
  buildErrorMessage,
  buildPartialMessage,
  isPcmFrame,
  parseStartMessage,
  type Prediction,
} from "../src/protocol.ts";

describe("protocol schemas", () => {
  test("accepts only the exact start message", () => {
    expect(parseStartMessage('{"type":"start"}').success).toBe(true);
    expect(parseStartMessage('{"type":"start","extra":true}').success).toBe(false);
    expect(parseStartMessage('{"type":"partial"}').success).toBe(false);
    expect(parseStartMessage("not-json").success).toBe(false);
  });

  test("validates the fixed 640 byte PCM frame size", () => {
    expect(isPcmFrame(new Uint8Array(FRAME_BYTES))).toBe(true);
    expect(isPcmFrame(new Uint8Array(FRAME_BYTES - 1))).toBe(false);
    expect(isPcmFrame("text")).toBe(false);
  });

  test("serializes partial messages with the loadgen schema fields", () => {
    const prediction: Prediction = {
      rms: 2425.5,
      zero_crossings: 7,
      checksum: 1745339397,
      samples: 1280,
      transcript: "now",
      audio_bytes: 640,
    };

    const payload = JSON.parse(
      buildPartialMessage({
        oldestFrameSeq: 1,
        newestFrameSeq: 2,
        frames: 2,
        cpuPasses: 4,
        modelDelayMs: 75,
        flushLatenessMs: 1.23456,
        prediction,
      }),
    );

    expect(payload).toEqual({
      type: "partial",
      oldest_frame_seq: 1,
      newest_frame_seq: 2,
      frames: 2,
      rms: 2425.5,
      zero_crossings: 7,
      checksum: 1745339397,
      samples: 1280,
      transcript: "now",
      audio_bytes: 640,
      cpu_passes: 4,
      model_delay_ms: 75,
      flush_lateness_ms: 1.235,
      inflight_model_jobs: 0,
    });
  });

  test("serializes inference errors with the diagnostic schema fields", () => {
    const payload = JSON.parse(
      buildErrorMessage({
        stage: "inference_request",
        kind: "timeout",
        message: "timed out",
        oldestFrameSeq: 1,
        newestFrameSeq: 2,
        frames: 2,
        audioBytes: 1280,
        oldestAgeMs: 1200.12345,
        newestAgeMs: 200.12345,
        flushLatenessMs: 4.4444,
        inferenceElapsedMs: 2000.12345,
        inflightGatewayBatches: 1,
        gatewayBufferFrames: 3,
        inferenceStatus: null,
        retryable: true,
      }),
    );

    expect(payload).toEqual({
      type: "error",
      stage: "inference_request",
      kind: "timeout",
      message: "timed out",
      oldest_frame_seq: 1,
      newest_frame_seq: 2,
      frames: 2,
      audio_bytes: 1280,
      oldest_age_ms: 1200.123,
      newest_age_ms: 200.123,
      flush_lateness_ms: 4.444,
      inference_elapsed_ms: 2000.123,
      inflight_gateway_batches: 1,
      gateway_buffer_frames: 3,
      inference_status: null,
      retryable: true,
    });
  });
});
